vector-im/element-android
# Improve wellknown usage

This PR fixes several issues:

Fixes #2843 : we now first perform a .well-known request, even if the entered URL is a valid Matrix API

Fixes #3572: we now store the url entered at session creation (Auth database migration required to add the new field), for further reference. Settings screen is updated:

<img width="312" alt="image" src="https://user-images.githubusercontent.com/3940906/123944375-5b501580-d99d-11eb-8ea2-2d230a20cb4f.png">

and

<img width="329" alt="image" src="https://user-images.githubusercontent.com/3940906/123944430-66a34100-d99d-11eb-831f-0bef2444f448.png">

API URL is only displayed in developer mode (see the diff with https://user-images.githubusercontent.com/3940906/123458079-e79cba00-d5e4-11eb-825e-4d481662b418.png).

Some change in the SDK API, wellknown now takes a domain instead of a matrixId.

### Tested OK

- sign in on matrix.org
- sign up on matrix.org
- SSO on matrix.org (using a GitHub account, see the screenshots)
- sign in on mozilla.modular.im (other SSO)
- sign in using a matrixId
- use element web url (develop.element.io) to retrieve config file

### Limitation

If the user enter the matrix API url, for instance "matrix-client.matrix.org", there is no way to deduce the domain matrix.org. So every further wellknow request will fail.
